### ISSUE TITLE
Error in Log.js when logging data.data

### ISSUE THREAD

**Issue by deadratfink** [Unitech/pm2#1765]
Hi,

we just updated from v0.14.7 to to v0.15.10 of PM2. We experienced an error in Log.js line 97:

```javascript
 {"name":"bdu-node-rest-service-template(development)","hostname":"HBV83032667","pid":3183,"level":30,"msg":"process.env.NEW_RELIC_HOME = ....","time":"2015-11-10T13:11:21.548Z","v":0}
DEBUUUUUUUUUG: {"at":1447161081557,"data":{"type":"Buffer","data":[123,34,118,34,58,48,44,34,108,101,118,101,108,34,58,50,48,44,34,110,97,109,101,34,58,34,110,101,119,114,101,108,105,99,34,44,34,104,111,115,116,110,97,109,101,34,58,34,72,66,86,56,51,48,51,50,54,54,55,34,44,34,112,105,100,34,58,51,49,56,51,44,34,116,105,109,101,34,58,34,50,48,49,53,45,49,49,45,49,48,84,49,51,58,49,49,58,50,49,46,53,53,54,90,34,44,34,109,115,103,34,58,34,76,111,97,100,105,110,103,32,97,103,101,110,116,32,102,114,111,109,32,47,85,115,101,114,115,47,106,101,110,115,46,107,114,101,102,101,108,100,116,47,68,101,118,101,108,111,112,109,101,110,116,47,83,111,117,114,99,101,115,47,110,111,100,101,106,115,47,98,100,117,45,110,111,100,101,45,114,101,115,116,45,115,101,114,118,105,99,101,45,116,101,109,112,108,97,116,101,47,110,111,100,101,95,109,111,100,117,108,101,115,47,98,100,117,45,110,111,100,101,45,110,101,119,114,101,108,105,99,47,110,111,100,101,95,109,111,100,117,108,101,115,47,110,101,119,114,101,108,105,99,34,125,10]},"process":{"pm_id":0,"name":"bdu-node-rest-service-template-1.1.0","rev":"392b26a76ca7fa186b2dc51674fab5884aae938c"}} data.data="[object Object]"
/Users/jens.krefeldt/Development/Sources/nodejs/bdu-node-rest-service-template/node_modules/bdu-node-pm2/node_modules/pm2/lib/Log.js:98
      var lines = (data.data || '').split('\n');
                                    ^

TypeError: (data.data || "").split is not a function
    at Parser.<anonymous> (/Users/jens.krefeldt/Development/Sources/nodejs/bdu-node-rest-service-template/node_modules/bdu-node-pm2/node_modules/pm2/lib/Log.js:98:37)
    at Parser.<anonymous> (/Users/jens.krefeldt/Development/Sources/nodejs/bdu-node-rest-service-template/node_modules/bdu-node-pm2/node_modules/pm2/node_modules/pm2-axon/lib/sockets/sub-emitter.js:52:19)
    at emitOne (events.js:77:13)
    at Parser.emit (events.js:169:7)
    at Parser._write (/Users/jens.krefeldt/Development/Sources/nodejs/bdu-node-rest-service-template/node_modules/bdu-node-pm2/node_modules/pm2/node_modules/pm2-axon/node_modules/amp/lib/stream.js:91:16)
    at doWrite (_stream_writable.js:292:12)
    at writeOrBuffer (_stream_writable.js:278:5)
    at Parser.Writable.write (_stream_writable.js:207:11)
    at Socket.ondata (_stream_readable.js:528:20)
    at emitOne (events.js:77:13)
```

I inserted a `console.log` before line 97 (so it is shown as line 98 above, so don't be confused ;-):

```javascript
      //console.log('DEBUUUUUUUUUG: ' + JSON.stringify(data) + ' data.data=' + JSON.stringify(data.data.toString()));
      var lines = (data.data || '').split('\n');  
```

which shows me that the incoming `data.data` is set but is NOT a string but a an object which itself contains a `data.data`, so it's actually a `data.data.data`:

```
DEBUUUUUUUUUG: {"at":1447161081557,"data":{"type":"Buffer","data":[123,34,118,34,58,48,44,34,108,101,118,101,108,34,58,50,48,44,34,110,97,109,101,34,58,34,110,101,119,114,101,108,105,99,34,44,34,104,111,115,116,110,97,109,101,34,58,34,72,66,86,56,51,48,51,50,54,54,55,34,44,34,112,105,100,34,58,51,49,56,51,44,34,116,105,109,101,34,58,34,50,48,49,53,45,49,49,45,49,48,84,49,51,58,49,49,58,50,49,46,53,53,54,90,34,44,34,109,115,103,34,58,34,76,111,97,100,105,110,103,32,97,103,101,110,116,32,102,114,111,109,32,47,85,115,101,114,115,47,106,101,110,115,46,107,114,101,102,101,108,100,116,47,68,101,118,101,108,111,112,109,101,110,116,47,83,111,117,114,99,101,115,47,110,111,100,101,106,115,47,98,100,117,45,110,111,100,101,45,114,101,115,116,45,115,101,114,118,105,99,101,45,116,101,109,112,108,97,116,101,47,110,111,100,101,95,109,111,100,117,108,101,115,47,98,100,117,45,110,111,100,101,45,110,101,119,114,101,108,105,99,47,110,111,100,101,95,109,111,100,117,108,101,115,47,110,101,119,114,101,108,105,99,34,125,10]},"process":{"pm_id":0,"name":"bdu-node-rest-service-template-1.1.0","rev":"392b26a76ca7fa186b2dc51674fab5884aae938c"}} data.data="[object Object]"
```

Since we use _newrelic_ module it seems this strange object which I printed out is in the context of it because the error occurs right after the _newrelic_ plugin was added by `require('newrelic')` at the top of Node.js script _www_ which is loaded by PM2 programmatically (using a `pm2.start({...}, cb)` in the callback of `pm2.connect(cb)`, ...you know). 

Another hint for that is that when I place the `require('newrelic')` inside the the `cb` of `pm2.start({...}, cb)` the error disappear.

Anyway, I don't know why the latter happens, but we have the issue here that some other code could provoke such objects coming to line 97 which are _not_ a string (but which is assumed by this line)!

I assume that for performance reasons there is not a `typeof` check, but for the sake of more robustness I would propose to do a check or use even use a `JSON.stringify(data.data)` to prevent such situations. Unfortunately, this issue  prevents us from using the newest version. If you need more information pls let me know...

What do you think?

Thx, Jens

**Comment by Unitech** [comment]
Hello @deadratfink, thanks for this bug report, could you please provide us a simple gist to reproduce the bug? 

Thanks

**Comment by robhawkes** [comment]
TL;DR: Looks like pm2 was getting confused between 2 different installed versions – removing all traces of pm2 from my system and starting afresh prevented these errors.

I was getting the same issue as reported in here and #1752 and noticed that I was also seeing the following pm2 version error, even though I was installing the latest pm2:

```bash
>>>> In-memory PM2 is out-of-date, do:
>>>> $ pm2 updatePM2
In memory PM2 version: 0.12.16
Local PM2 version: 0.15.10
```

It turns out that a previous version of pm2 was still running even though I had installed the latest version and run the `pm2 updatePM2` command. Nothing I seemed to do would get the `pm2` command to recognise the latest version.

What I realised was that running `which pm2` returned a directory that clearly wasn't related to where modules are installed now that I use nvm, so it must have been a super old version I had installed before moving to nvm. Once I realised that I switched to the system Node version using nvm and re-ran all the uninstall commands for pm2, including running `pm2 kill` – only then did `which pm2` return nothing.

From that point I switched to the latest version of Node using nvm and reinstalled pm2 – everything worked and no more `Log.js` errors.

Maybe this helps, or at least narrows down a cause.

**Comment by deadratfink** [comment]
Hi,

Thx for that hint, I will try this after my vacation will have ended (i.e. beginning of Jan 2016) and retry the newest version on our services. Maybe I had this version mixup as well (for whatever reasons, at least I cannot say if I had such kind of trouble like robhawkes reported but I would not exclude such possibility).

**Comment by AdrienFromToulouse** [comment]
Hello all,

I was running my PM2 processes as follow (in a Docker container: `ubuntu 14`, `node 4.x`, `pm2 0.15.7`):

```CMD pm2 start --no-daemon  processes.json```

Inside `processes.json` I had an app in `cluster_mode`.
It appears that it was the root of the issue. I do not have that issue in `fork mode`...

Indeed it was the app in cluster mode which was failing to log.

Hope it could help somehow. Cheers!

**Comment by nithin-bose** [comment]
I have the same issue. the app runs fine in fork mode but fails in cluster mode. I am running in docker with the -i param set to 0

**Comment by aunz** [comment]
Not sure if this helps
The snippet below can produce the `TypeError: (data.data || "").split is not a function`
On ubuntu 14.04 64-bit, node 5.7.0, PM2 1.0.2


```js
// app.js
require('http').createServer((req, res) => {
  res.writeHead(200, { 'Content-Type': 'text/plain' });
  res.end('Hello World\n');
  require('child_process').execSync('ls') // this line causes the error
  // require('child_process').exec('ls', (e, r) => {}) // this line doesn't cause the error
}).listen(8080, () => {});
```
Only in `pm2 start app.js -i 0`
NOT in `pm2 start app.js`
So, a similar observation mentioned by @AdrienFromToulouse but only when a child_process is executed in sync mode

**Comment by Unitech** [comment]
@aunz thanks bug reproduced! Will fix that by today

**Comment by hbakhtiyor** [comment]
any info about this issue? what's the plan?

**Comment by Unitech** [comment]
Fix on his way, https://github.com/Unitech/pm2/blob/development/lib/Log.js#L100

Just need to do some checkings (memory, cpu profiling etc) before publishing the new pm2 version.

In the meantime you can use it:

```
$ npm install Unitech/pm2#development -g
```
